devices/HumanStateProvider/HumanStateProvider.cpp

```diff
@@ -654,6 +670,221 @@ bool HumanStateProvider::open(yarp::os::Searchable& config)
+            }
+        } // another option
+        else {
+            yError() << LogPrefix << "the parameter key is not defined: " << constraintKey;
+            return false;
+        }
+    }
+
+    // check sizes
+    if (pImpl->custom_jointsVelocityLimitsNames.size()
+        != pImpl->custom_jointsVelocityLimitsValues.size()) {
+        yError() << "the joint velocity limits name and value size are not equal";
+        return false;
+    }
+    if ((pImpl->customConstraintUpperBound.size() != pImpl->customConstraintLowerBound.size())
+        && (pImpl->customConstraintLowerBound.size() != pImpl->customConstraintMatrix.rows())) {
+        yError() << "the number of lower bound (" << pImpl->customConstraintLowerBound.size()
+                 << "), upper buond(" << pImpl->customConstraintUpperBound.size()
+                 << "), and cosntraint matrix rows(" << pImpl->customConstraintMatrix.rows()
+                 << ") are not equal";
+
+        return false;
+    }
+    if ((pImpl->customConstraintVariables.size() != pImpl->customConstraintMatrix.cols())) {
+        yError() << "the number of constraint variables ("
+                 << pImpl->customConstraintVariables.size() << "), and cosntraint matrix columns ("
+                 << pImpl->customConstraintMatrix.cols() << ") are not equal";
+        return false;
+    }
+    yInfo() << "******* DOF: " << modelLoader.model().getNrOfDOFs();
+    for (size_t i = 0; i < pImpl->custom_jointsVelocityLimitsNames.size(); i++) {
+        pImpl->custom_jointsVelocityLimitsIndexes.push_back(
+            modelLoader.model().getJointIndex(pImpl->custom_jointsVelocityLimitsNames[i]));
+        yInfo() << pImpl->custom_jointsVelocityLimitsNames[i] << " : "
+                << pImpl->custom_jointsVelocityLimitsIndexes[i];
+    }
+    for (size_t i = 0; i < pImpl->customConstraintVariables.size(); i++) {
+        pImpl->customConstraintVariablesIndex.push_back(
+            modelLoader.model().getJointIndex(pImpl->customConstraintVariables[i]));
+        yInfo() << pImpl->customConstraintVariables[i] << " : "
+                << pImpl->customConstraintVariablesIndex[i];
+    }
+
+    //******************************************
+    //****** SET CONSTRAINTS FOR IB-IK *********
+    //******************************************
+    if (pImpl->custom_jointsVelocityLimitsNames.size() != 0) {
+        pImpl->inverseVelocityKinematics.setCustomJointsVelocityLimit(
+            pImpl->custom_jointsVelocityLimitsIndexes, pImpl->custom_jointsVelocityLimitsValues);
+    }
+    if (pImpl->baseVelocityUpperLimit.size() != 0) {
+        pImpl->inverseVelocityKinematics.setCustomBaseVelocityLimit(pImpl->baseVelocityLowerLimit,
+                                                                    pImpl->baseVelocityUpperLimit);
+    }
+    if (pImpl->customConstraintVariablesIndex.size() != 0) {
+        pImpl->inverseVelocityKinematics.setCustomConstraintsJointsValues(
```

Comment:
Here you are setting all the joint constraints passed trough configuration, however the model joint constraints are still missing, and are still enforced just by saturating the value of the joints during the integration step (see https://github.com/robotology/human-dynamics-estimation/blob/feature/AddQP_rebased/devices/HumanStateProvider/HumanStateProvider.cpp#L1444). We should instead add those limits also in the custom constraints matrix so that are taken into account in the QP, and eventually leave them also as saturation in order to be sure the values are never reached.

Reply:
To track the discussion for this comment, I opened this issue
https://github.com/robotology/human-dynamics-estimation/issues/150
I have added the functionality, however I think it is better to keep also the hard saturation constraints because of this comment: https://github.com/robotology/human-dynamics-estimation/issues/132#issuecomment-511417846

Reply:
> I have added the functionality, however I think it is better to keep also the hard saturation constraints because of this comment: [#132 (comment)](https://github.com/robotology/human-dynamics-estimation/issues/132#issuecomment-511417846)

I agree on keeping also the saturation